ec2/elasticip.py

```diff
@@ -14,109 +14,168 @@
+
+    address = utils.get_address_object_by_id(address_object.public_ip, ctx=ctx)
+
+    if not address:
+        elasticip = \
+            ctx.instance.runtime_properties.pop('aws_resource_id', None)
+        if 'allocation_id' in ctx.instance.runtime_properties:
+            del(ctx.instance.runtime_properties['allocation_id'])
+        ctx.logger.info(
+            'Released elastic ip {0}. '
+            'Removed runtime properties.'.format(elasticip))
+    else:
+        raise RecoverableError(
+            'Elastic IP not provisioned. Retrying...',
+            retry_after=retry_interval)
 
 
 @operation
-def associate(**kwargs):
+def associate(**_):
     """ Associates an Elastic IP with an EC2 Instance.
     """
+
+    ctx.logger.debug('Attempting to associate elasticip with instance.')
+
     ec2_client = connection.EC2ConnectionClient().client()
 
-    elasticip = ctx.target.node.properties['elasticip']
-    instance_id = ctx.source.instance.runtime_properties['instance_id']
-    ctx.logger.info('Associating an Elastic IP {0} '
-                    'with an EC2 Instance {1}.'.format(elasticip, instance_id))
+    if 'aws_resource_id' not in ctx.source.instance.runtime_properties:
+        raise NonRecoverableError(
+            'Unable to associate elastic ip address, '
+            'instance_id runtime property not set.')
+
+    if 'aws_resource_id' not in ctx.target.instance.runtime_properties:
+        raise NonRecoverableError(
+            'Unable to associate elastic ip address, '
+            'aws_resource_id runtime property not set.')
+
+    elasticip = ctx.target.instance.runtime_properties['aws_resource_id']
+    instance_id = ctx.source.instance.runtime_properties['aws_resource_id']
+
+    ctx.logger.debug('associate using elasticip {0} and instance {1}.'
+                     .format(elasticip, instance_id))
 
     try:
         ec2_client.associate_address(instance_id=instance_id,
                                      public_ip=elasticip)
-    except (EC2ResponseError, BotoServerError) as e:
-        raise NonRecoverableError('Error. Failed to '
-                                  'attach Elastic IP. Error: {0}.'
-                                  .format(e))
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('{0}.'.format(str(e)))
 
     ctx.logger.info('Associated Elastic IP {0} with instance {1}.'.format(
         elasticip, instance_id))
+    ctx.source.instance.runtime_properties['public_ip_address'] = elasticip
 
 
 @operation
-def disassociate(**kwargs):
+def disassociate(**_):
     """ Disassociates an Elastic IP from an EC2 Instance.
     """
     ec2_client = connection.EC2ConnectionClient().client()
-    elasticip = ctx.target.node.properties['elasticip']
+
+    if 'aws_resource_id' not in ctx.target.instance.runtime_properties:
+        raise NonRecoverableError(
+            'Failed to disossiate elastic ip, '
```

Comment:
discuss with ran if external_resource should have relationship operations performed on them or not.